pkg/networkservice/common/heal/eventloop.go

```diff
@@ -66,69 +68,78 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
 		eventLoopCancel()
-		return nil, errors.Wrap(err, "failed get MonitorConnections client")
+		return nil, nil, errors.Wrap(err, "failed get MonitorConnections client")
 	}
 
 	// get the initial state transfer and use it to detect whether we have a real connection or not
 	_, err = client.Recv()
 	if err != nil {
 		eventLoopCancel()
-		return nil, errors.Wrap(err, "failed to get the initial state transfer")
+		return nil, nil, errors.Wrap(err, "failed to get the initial state transfer")
 	}
 
 	logger := log.FromContext(ctx).WithField("heal", "eventLoop")
 	cev := &eventLoop{
-		heal:         heal,
-		eventLoopCtx: eventLoopCtx,
-		chainCtx:     ctx,
-		conn:         conn,
-		eventFactory: ev,
-		client:       newClientFilter(client, conn, logger),
-		logger:       logger,
+		heal:              heal,
+		eventLoopCtx:      eventLoopCtx,
+		eventLoopCancel:   eventLoopCancel,
+		chainCtx:          ctx,
+		conn:              conn,
+		eventFactory:      ev,
+		client:            newClientFilter(client, conn, logger),
+		logger:            logger,
+		monitorFinishedCh: make(chan bool, 1),
 	}
 
 	// Start the eventLoop
 	go cev.eventLoop()
-	return eventLoopCancel, nil
+	return eventLoopCancel, cev.monitorFinishedCh, nil
 }
 
-func (cev *eventLoop) monitorCtrlPlane() <-chan struct{} {
-	res := make(chan struct{}, 1)
-
-	go func() {
-		defer close(res)
+func (cev *eventLoop) eventLoop() {
+	needToHeal, reselect := cev.waitForEvents()
 
-		for {
-			eventIn, err := cev.client.Recv()
+	if !needToHeal {
+		return
+	}
 
-			if cev.chainCtx.Err() != nil || cev.eventLoopCtx.Err() != nil {
-				res <- struct{}{}
+	for {
+		select {
+		case <-cev.chainCtx.Done():
+			return
+		default:
+			if cev.chainCtx.Err() != nil {
 				return
 			}
 
-			// Handle error
-			if err != nil {
-				s, _ := status.FromError(err)
-				// This condition means, that the client closed the connection. Stop healing
-				if s.Code() == codes.Canceled {
-					res <- struct{}{}
+			// We need to force check the DataPlane if a down event was received from the ControlPlane
+			if !reselect {
+				deadlineCtx, deadlineCancel := context.WithDeadline(cev.chainCtx, time.Now().Add(cev.heal.livenessCheckTimeout))
+				if !cev.heal.livenessCheck(deadlineCtx, cev.conn) {
+					cev.logger.Warnf("Data plane is down")
+					reselect = true
 				}
-				// Otherwise - Start healing
-				return
+				deadlineCancel()
 			}
 
-			// Handle event. Start healing
-			if eventIn.GetConnections()[cev.conn.GetId()].GetState() == networkservice.State_DOWN {
+			var options []begin.Option
+			if reselect {
+				cev.logger.Debugf("Reconnect with reselect")
+				options = append(options, begin.WithReselect())
+			}
+			err := <-cev.eventFactory.Request(options...)
+			if err == nil {
+				cev.logger.Info("Heal success")
 				return
 			}
 		}
-	}()
-
-	return res
+	}
 }
 
-func (cev *eventLoop) eventLoop() {
-	reselect := false
+func (cev *eventLoop) waitForEvents() (needToHeal, reselect bool) {
```

Comment:
As I can see always if needToHeal is true then reselect is also true. 

Consider to merge to one bool response.